src/cgutils.cpp

```diff
@@ -337,66 +343,103 @@ JL_DLLEXPORT Type *julia_type_to_llvm(jl_value_t *jt, bool *isboxed)
     // use this where C-compatible (unboxed) structs are desired
     // use julia_type_to_llvm directly when you want to preserve Julia's type semantics
-    bool isTuple = jl_is_tuple_type(jt);
     if (isboxed) *isboxed = false;
-    if ((isTuple || jl_is_structtype(jt)) && !jl_is_array_type(jt)) {
+    if ((jl_is_tuple_type(jt) || jl_is_structtype(jt)) && !jl_is_array_type(jt)) {
         if (!jl_is_leaf_type(jt))
-            return NULL;
+            return nullptr;
         jl_datatype_t *jst = (jl_datatype_t*)jt;
-        if (jst->struct_decl == NULL) {
-            size_t ntypes = jl_datatype_nfields(jst);
-            if (ntypes == 0 || jst->size == 0)
+        if (jst->struct_decl == nullptr) {
+            assert(jl_datatype_nfields(jst) != 0 || jst->size == 0);
+            if (jst->size == 0)
                 return T_void;
-            StructType *structdecl;
-            if (!isTuple) {
-                structdecl = StructType::create(jl_LLVMContext, jl_symbol_name(jst->name->name));
-                jst->struct_decl = structdecl;
-            }
-            std::vector<Type*> latypes(0);
-            size_t i;
-            bool isvector = true;
-            Type *lasttype = NULL;
-            for(i = 0; i < ntypes; i++) {
-                jl_value_t *ty = jl_svecref(jst->types, i);
-                Type *lty;
-                if (jl_field_isptr(jst, i))
-                    lty = T_pjlvalue;
-                else
-                    lty = ty==(jl_value_t*)jl_bool_type ? T_int8 : julia_type_to_llvm(ty);
-                if (lasttype != NULL && lasttype != lty)
-                    isvector = false;
-                lasttype = lty;
-                if (type_is_ghost(lty))
-                    lty = NoopType;
-                latypes.push_back(lty);
-            }
-            if (!isTuple) {
-                structdecl->setBody(latypes);
-            }
-            else {
-                if (isvector && lasttype != T_int1 && !type_is_ghost(lasttype)) {
-                    // TODO: currently we get LLVM assertion failures for other vector sizes
-                    bool validVectorSize = (ntypes == 2 || ntypes == 4 || ntypes == 6);
-                    if (0 && lasttype->isSingleValueType() && !lasttype->isVectorTy() && validVectorSize) // currently disabled due to load/store alignment issues
-                        jst->struct_decl = VectorType::get(lasttype, ntypes);
-                    else
-                        jst->struct_decl = ArrayType::get(lasttype, ntypes);
-                }
-                else {
-                    jst->struct_decl = StructType::get(jl_LLVMContext,ArrayRef<Type*>(&latypes[0],ntypes));
-                }
-            }
+            set_struct_decl(jst);
         }
         return (Type*)jst->struct_decl;
     }
     return julia_type_to_llvm(jt, isboxed);
 }
 
+// Returns 0 if no special rule applies
+unsigned jl_llvm_special_tuple_alignment(jl_tupletype_t *st)
```

Comment:
The tricky part is that currently the tuple type is used if all the elements have the same *LLVM* type and the LLVM type is a scalar type.  I'll have to cut that back to using a tuple only if all the elements have the same Julia type, or all are pointers.  And make an estimate on whether the Julia element type maps to an LLVM scalar type (seems straightforward).

I suspect the pointer case isn't interesting, though AVX-512 and a creative mind will prove me wrong.